File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/observability/otel_span_sink.rb
# frozen_string_literal: true

require "monitor"

module IAIso
  module Observability
    # IAIso OpenTelemetry tracing sink.
    #
    # Duck-typed against the OTel trace API. The official
    # `opentelemetry-api` gem's tracer / span objects satisfy these
    # contracts with thin adapters.
    #
    # Required interfaces (duck typing):
    #   tracer.start_span(name, attributes:) -> span
    #   span.add_event(name, attributes: ...)
    #   span.set_attribute(key, value)
    #   span.finish
    class OtelSpanSink
      include MonitorMixin

      def initialize(tracer:, span_name: "iaiso.execution")
        super()
        @tracer = tracer
        @span_name = span_name
        @spans = {}
      end

      def emit(event)
        synchronize do
          if event.kind == "engine.init"
            @spans[event.execution_id] = @tracer.start_span(
              "#{@span_name}:#{event.execution_id}",
              attributes: { "iaiso.execution_id" => event.execution_id },
            )
          end
          span = @spans[event.execution_id]
          return if span.nil?

          attrs = event.data.merge("iaiso.schema_version" => event.schema_version)
          span.add_event(event.kind, attributes: attrs)
          case event.kind
          when "engine.step"
            span.set_attribute("iaiso.pressure", event.data["pressure"]) if event.data["pressure"]
          when "engine.escalation"
            span.set_attribute("iaiso.escalated", true)
          when "engine.release"
            span.set_attribute("iaiso.released", true)
          when "execution.closed"
            span.finish
            @spans.delete(event.execution_id)
          end
        end
      end

      def close_all
        synchronize do
          @spans.each_value { |s| s.finish rescue nil }
          @spans.clear
        end
      end
    end
  end
end

File: IAIso-v5.0/core/iaiso-java/iaiso-observability/src/main/java/io/iaiso/observability/OtelSpanSink.java
package io.iaiso.observability;

import io.iaiso.audit.Event;
import io.iaiso.audit.Sink;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * IAIso OpenTelemetry tracing sink.
 *
 * <p>Structurally typed against the OTel trace API. The official
 * {@code opentelemetry-java} {@code Tracer} and {@code Span} satisfy
 * these interfaces with thin adapters.
 */
public final class OtelSpanSink implements Sink {

    public interface Span {
        void addEvent(String name, Map<String, Object> attributes);
        void setAttribute(String key, Object value);
        void end();
    }

    public interface Tracer {
        Span startSpan(String name, Map<String, Object> attributes);
    }

    private final Tracer tracer;
    private final String spanName;
    private final ConcurrentHashMap<String, Span> spans = new ConcurrentHashMap<>();

    public OtelSpanSink(Tracer tracer, String spanName) {
        this.tracer = tracer;
        this.spanName = (spanName == null || spanName.isEmpty())
            ? "iaiso.execution" : spanName;
    }

    /** End any open spans. Useful at shutdown. */
    public void closeAll() {
        for (Span s : spans.values()) {
            try { s.end(); } catch (RuntimeException ignored) {}
        }
        spans.clear();
    }

    @Override
    public void emit(Event event) {
        Span span;
        if ("engine.init".equals(event.getKind())) {
            Map<String, Object> attrs = new HashMap<>();
            attrs.put("iaiso.execution_id", event.getExecutionId());
            span = tracer.startSpan(spanName + ":" + event.getExecutionId(), attrs);
            spans.put(event.getExecutionId(), span);
        } else {
            span = spans.get(event.getExecutionId());
        }
        if (span == null) return;

        Map<String, Object> attrs = new HashMap<>(event.getData());
        attrs.put("iaiso.schema_version", event.getSchemaVersion());
        span.addEvent(event.getKind(), attrs);

        switch (event.getKind()) {
            case "engine.step":
                Object p = event.getData().get("pressure");
                if (p != null) span.setAttribute("iaiso.pressure", p);
                break;
            case "engine.escalation":
                span.setAttribute("iaiso.escalated", true);
                break;
            case "engine.release":
                span.setAttribute("iaiso.released", true);
                break;
            case "execution.closed":
                span.end();
                spans.remove(event.getExecutionId());
                break;
            default:
                // ignore
        }
    }
}
